### ISSUE TITLE
On web, casting null as non-nullable T succeeds when T is passed from explicit extension incovation

### ISSUE THREAD

**Issue by LewisHolliday** [dart-lang/sdk#54514]
Example:
```dart
extension E<T> on Object {
  T foo() {
    return null as T;
  }
}

void main() {
  final Object a = E<Object>(0).foo();
  print(a); // null
  
  final String b = E<String>(0).foo();
  print(b); // null
}
```
Tested on DartPad and `dart compile js`.

**Comment by eernstg** [comment]
Looks like the actual type argument in the explicit extension method invocation is ignored at run time. Apparently, it occurs only with web targets&mdash;`dart` and `dart compile exe` do perform the type casts and raise the run-time error, but both DartPad (and `dart compile js`, as reported above) have the soundness issue.

**Comment by LewisHolliday** [comment]
It seems that this issue also happens with non-extension methods:
```dart
T foo<T>() => null as T;
void main() => print(foo<int>()); // null
```

**Comment by eernstg** [comment]
Weird!

```dart
T foo<T>() {
  print(T);
  return null as T; // Throws, as it should.
}

void main() => print(foo<int>());
```

**Comment by sigmundch** [comment]
DDC works properly, this failure only occurs in dart2js. Could be related to bugs in the rti-needs optimization.

**Comment by sigmundch** [comment]
Indeed: the last example that explicitly uses `T` in the print lead me to believe this was related to the rti-needs. And indeed, using `--disable-rti-optimization` also fixes the issue.

**Comment by fishythefish** [comment]
This should be fixed by https://dart-review.googlesource.com/c/sdk/+/343821.
